ci/daily_tell_slack.yml

```
@@ -16,7 +16,7 @@ steps:
         MESSAGE=${{ parameters['success-message'] }}
     fi
     PAYLOAD="{\"text\":\"$MESSAGE\n\"}"
-    if [ "$(variables['Build.SourceBranchName'])" = "master" ]; then
+    if [ "$(Build.SourceBranchName)" = "master" ]; then
```

If you want to fail eagerly if this variable doesn't exist, you could add:

```suggestion
    [ -n "$(Build.SourceBranchName)" ]
    if [ "$(Build.SourceBranchName)" = "master" ]; then
```

---

Is `set -u` not enough for that?

---

Clearly `set -u` is not working here, as the jobs are still green on Azure despite the error message.

---

Nope, because this isn't a shell variable, it's an Azure Pipelines variable. It's inserted through string interpolation before the script is run, as far as I understand it.

Perhaps the entire job will fail if the variable doesn't exist though. Not sure how permissive Azure is with this stuff. A quick search suggests it will either insert an empty string or "undefined", which seems even less useful.

---

More layers of string interpolation solve all problems.

---

@SamirTalwar That is not how Azure works. Not sure what you searched for, but if the variable doesn't exist at the Azure level, or if the syntax is wrong (as was the case here), Azure simply does not replace anything. So in this case Bash did get the literal `"$(variables['SourceBranchName'])"` and that is why Bash (not Azure preprocessing) printed
```
/home/vsts/agent/_work/_temp/7a958ea0-e48d-4658-9200-9d933a35582a.sh: line 11: variables[Build.SourceBranchName]: command not found
```
from its subshell, of which the return status is ignored despite `-e` because we're inside a `[]` (I think).

So I don't think the `[ -n ]` will help here. From my understanding of `-e` (`-u` is irrelevant here as from Bash's perspective there is no variable involved, just a subshell, so nothing that can be unset), this should fail properly on the subshell failing:
```suggestion
    branch="$(Build.SourceBranchName)"
    if [ "$branch" = "master" ]; then
```
Either way, we now have the correct syntax so this will never fail (🤞), as the `Build.SourceBranchName` variable should always be set by Azure.

---

Thanks for the explanation @garyverhaegen-da!

---

Aha, good to know. Thanks for the clarification!